qadence/backends/utils.py

```diff
@@ -90,3 +95,78 @@ def to_list_of_dicts(param_values: dict[str, Tensor]) -> list[dict[str, float]]:
     }
 
     return [{k: v[i] for k, v in batched_values.items()} for i in range(max_batch_size)]
+
+
+def finitediff_sampling(
+    f: Callable, x: torch.Tensor, eps: float = FINITE_DIFF_EPS, num_samples: int = 10
+) -> torch.Tensor:
+    def _finitediff(val: torch.Tensor) -> torch.Tensor:
+        return (f(x + val) - f(x - val)) / (2 * val)  # type: ignore
+
+    with torch.no_grad():
+        return torch.mean(
+            torch.cat([_finitediff(val) for val in torch.rand(1) for _ in range(num_samples)])
+        )
+
+
+def finitediff(f: Callable, x: torch.Tensor, eps: float = FINITE_DIFF_EPS) -> torch.Tensor:
+    return (f(x + eps) - f(x - eps)) / (2 * eps)  # type: ignore
+
+
+def dydx(
```

Comment:
@madagra thats how we could do higher order stuff. we will have to check numerical stability /quality though

Reply:
Are these functions used somewhere in the code? If not, I would mention that in a comment since they will be used later on, I think.

Reply:
done